### ISSUE TITLE
[BUG] OIIO seems to consider that the alpha channel is gamma-compressed when reading/writing 8-bit formats

### ISSUE THREAD

**Issue by devernay** [OpenImageIO/oiio#2887]
**Describe the bug**

Use the attached images `ramp.png` and `solidblack.png`.
ramp.png has a linear alpha variation from left to right (8-bit values checked with the GIMP)
and do:
```
oiiotool ramp.png --ch "R,G,B" -o rampoiiorgb.png # result way too dark
oiiotool --no-autopremult ramp.png --tocolorspace linear --premult --ch "R,G,B" --tocolorspace sRGB -o rampoiioopaque.png # GOOD expected result
oiiotool ramp.png --ch "R,G,B" -o rampoiiorgb.exr # result not fully OK, the red and black bands are too narrow
oiiotool rampoiiorgb.exr --tocolorspace sRGB -o rampoiiorgbviaexr.png # same

# composite over solid black
oiiotool ramp.png solidblack.png -over -o rampoversolidblackoiio.png # result too dark
oiiotool --no-autopremult ramp.png --tocolorspace linear --premult solidblack.png -over --tocolorspace sRGB -o rampoversolisblackoiio2.png # GOOF expected result
oiiotool ramp.png solidblack.png -over -o rampoversolidblackoiio.exr # result not fully OK
oiiotool rampoversolidblackoiio.exr --tocolorspace sRGB -o rampoversolidblackoiioviaexr.png # result OK

```

I checked manually the values at x=960, y=540:
- ramp.png has RGBA value (177,0,196,128) (50% alpha, unassociated because PNG)
- rampoiiorgb.png and rampoversolidblackoiio.png have RGB value (89,0,98)
- rampopaque.png (attached image) has RGB value (129,0,143) (it was obtained by linearizing RGB from ramp.png, then compositing over black with the linear alpha or equivalently multiply by alpha, then delinearizing RGB)
- rampoiioopaque.png and rampoversolisblackoiio2.png have RGB value (129,0,144), close to what I compute by hand
- rampoiiorgbviaexr.png and rampoversolidblackoiioviaexr.png have RGB value (159,0,167), which I cannot explain, where do these values come from?
- R_good = ((177/255)^2.2 * (128/255))^(1/2.2) * 255 = 129 <-- I should get this value
- B_good = ((196/255)^2.2 * (128/255))^(1/2.2) * 255 = 143 <-- id
- R_bad = ((177/255)^2.2 * (128/255)^2.2)^(1/2.2) * 255 = 89 <-- I get that instead: alpha was not considered linear it seems! Or alpha was applied on the nonlinear RGB: this is also simply 177 * (128/255) = 89
- B_bad = ((196/255)^2.2 * (128/255)^2.2)^(1/2.2) * 255 = 98 <-- id, this is 196 * (128/255) = 98

I see two or three potential bugs, or I may not have understood what happened:
1. converting via EXR should give the same result as to PNG
2. PNG result is too dark (darker than the EXR version)
3. Even the EXR version seems wrong with respect to what I compute

This may also be the root cause of bug https://github.com/NatronGitHub/Natron/issues/582

See also:
- [What every coder should know about gamma](https://blog.johnnovak.net/2016/09/21/what-every-coder-should-know-about-gamma/#colour-blending) which discusses the Photoshop option "Blend RGB colors using gamma 1" (which should be set by default, but is not because of 8-bit legacy)
- [[css3-images] We should use linear RGB color space when doing gradients (and transitions, and ...)](https://lists.w3.org/Archives/Public/www-style/2012Jan/0607.html)
- [Gamma error in picture scaling](http://www.ericbrasseur.org/gamma.html)
- https://discuss.pixls.us/t/ocio-srgb-alpha-conversion-problem/21040

**Platform information:**
 - OIIO branch/version: 2.2.7 (from Mac homebrew)
 - OS: macOS 11.2
 - C++ compiler: from Xcode
 - Any non-default build flags when you build OIIO: homebrew build

ramp.png:
![ramp](https://user-images.githubusercontent.com/3778071/110172144-8f317a80-7db1-11eb-869c-1566bde7fc46.png)

solidblack.png:
![solidblack](https://user-images.githubusercontent.com/3778071/110172147-9062a780-7db1-11eb-99c1-a814b3aaace8.png)

rampopaque.png:
![rampopaque](https://user-images.githubusercontent.com/3778071/110174860-a83c2a80-7db5-11eb-944d-c0c1c544c7c6.png)

**Comment by devernay** [comment]
BTW I disagree with this comment from the OIIO doc:

"PNG stupidly specifies that any alpha channel is “unassociated” (i.e., that the color channels are not “premulti-plied” by alpha). This is a disaster, since it results in bad loss of precision for alpha image compositing, and evenmakes it impossible to properly represent certain additive glows and other desirable pixel values."

Having *associated* alpha in an 8bit format would be a "disaster" and a "bad loss of precision". Unassociated alpha is the only way to have the correct colors on transparent pixels. The *only* downside is what is mentionned about glow: you cannot make a pixel "brighter than white", but this is also mostly true for integer-based formats, which have a limited range.

**Comment by lgritz** [comment]
Ugh, this makes my head hurt. We have both nonlinear input as well as unassociated alpha, and at least for me, that makes it very difficult to properly reason about what it going on.

I feel like the first thing to do is to make sure we understand how it's *supposed* to look and be using the same nomenclature.

For the sake of this discussion, let's use "png" to mean the original encoded values in the files: uint8, sRGB transfer function (but linear transfer function for alpha), and unassociated alpha (a.k.a. "non-premultiplied" colors). 

And let's use "true" to mean a linear response, float 0.0-1.0 normalized display-referred (that is 1.0 is the brightest the screen can get), with associated alpha values (a.k.a. premultiplied color).

The ramp.png file, at the leftmost edge (x=0) has png value (0, 255, 0, 0). What do you think its true value is, and when displayed, how do you think that should look?  I think the true value is (0.0, 0.0, 0.0, 0.0) and it looks black, NOT BRIGHT GREEN. Do you agree?

I would convert the "raw" png to a "true" exr like this:

    oiiotool --autocc ramp.png -o ramp.exr

(The --autocc means that every file, upon input, is automatically converted to a canonical linear color space based on the name and/or file format, and every output will be convert to any color space that is required or implied by its file type or naming convention. PNG is known to be sRGB, OpenEXR is known to be linear, unless file name convention overrides.)

OK, when you display ramp.exr in your favourite color-aware display program, does it look right?

I feel like we should stop here and make sure we are synchronized, before taking the next step. Because if we get that far agreeing that we have properly converted to a linear, associated-alpha exr, then we can use a parallel pipeline on the exr files to check that whatever we do involving the pngs is matching.

**Comment by devernay** [comment]
Hi Larry, I 100% agree it is very important to understand *if* there's something wrong going on, or if I'm misusing OIIO.

The left-most pixel can be any color, and that's OK, since it's transparent (that's the magic of unassociated alpha), but I guess green was something like 1e-8 and alpha was 1e-8, son Natron unpermiltiplied it to (0,1,0,1e-8), and then conversion to 8bit droppend the very small alpha.

I was looking more at the mid-range (x=960), because that's usually where you see things happening.

For me,
```
oiiotool --autocc ramp.png -o ramp.exr
```
gives exactly the same result as
```
oiiotool ramp.png -o ramp.exr
```
Is that expected? I would expect OIIO to have that kind of cleverness.

If I keep only the RGB channels (for visualization purposes, since my screen doesn't know how to be transparent) using `--ch "R,G,B"`, I get the same result as the "way too dark" examples on top, except in linear EXR instead of sRGB PNG.

What *seems* to be happening, and maybe I should have started by that is that OIIO *first* associates alpha on the original RGB values from the PNG, which are sRGB, and *then* linearizes them. This should be the other way round, of course, since any color conversion must be done on unassociated values. Would that be a potential explanation?

I get what I think is the right result by reading a PNG this way:
```
oiiotool --no-autopremult ramp.png --tocolorspace linear --premult -o rampgood.exr
# to check the value of premultiplied colors, same result as rampoiioopaque.png above, but linear
oiiotool rampgood.exr --ch 'R,G,B' -o rampgoodrgb.exr
```
And I can get back to the original PNG from the linear EXR this way:
```
oiiotool rampgood.exr --unpremult --tocolorspace sRGB -o rampgood.png
idiff -v rampgood.png ramp.png
Comparing "rampgood.png" and "ramp.png"
  Mean error = 0
  RMS error = 0
  Peak SNR = inf
  Max error  = 0
  0 pixels (0%) over 1e-06
  0 pixels (0%) over 1e-06
PASS
```
If I don't explicitly unpremult before conversion to sRGB, it seems like (pure guess) it is done *after* conversion to sRGB, and you can see the colors are too bright when you visualize it. pixel (960,540) has color values (255,0,255,128) - only alpha is right :-(
```
oiiotool rampgood.exr --tocolorspace sRGB -o rampsrgb.png
```

Maybe the main issue comes from the fact that a premultiplied workflow only makes sense in a linear colorspace? 

Maybe OIIO should not try to associate alpha when the colorspace is not linear?

Or is it just a bug where the premult/unpremult happens at the wrong place when reading/wrinting sRGB images?

**Comment by lgritz** [comment]
Aaah, I think you have put your finger on the real problem.

Let's say we have a raw sRGB unpremultipled (255, 0, 0, 128), which is equivalent to normalized (1.0, 0.0, 0.0, 0.5).   [Note: to make this math easy to follow with round numbers, I'm going to pretend that 128 is 0.5 when normalized, and also that sRGB is gamma 2.0 -- I know it's not!]

If you premultiply first, you get (0.5, 0.0, 0.0, 0.5), and then linearize (which we simplify to pow(x,2)) to get (0.25, 0.0, 0.0, 0.5), a dark red.

If we linearize first, we get (1.0, 0.0, 0.0, 0.5), because 1.0 stays 1.0 under linearization, and then premultiply to get (0.5, 0.0, 0.0, 0.5). This is more of a medium red.

I believe the second choice (linearize first) must be correct, because the original unpremult (255, 0, 0, 128) is supposed to be equivalent to a fully bright red covering 1/2 the pixel. And that was what we got in our second results, but not the first.

So I agree with you that the color correction / linearization should happen before the premultiplication, otherwise things can look too dark. And I do think you are correct that it's currently done the other way around, because the PNG reader itself is doing the premult (unless instructed not to) before the autocc logic in oiiotool kicks in and does the color transformation.

Let me prepare a PR that fixes this, and see if we both agree that it's giving the right results.

**Comment by lgritz** [comment]
As an aside, I think I have always had trouble understanding why the color conversion is typically implemented by dividing by alpha, doing the color space transform, and then re-multiplying by alpha. I had trouble wrapping my mind around why that was more correct that directly color correcting the premultiplied color. I've talked to color scientists about this and never got an explanation that I fully grokked, never quite agreed, but just decided they must know better than me. This example finally makes me understand in a simple way that I can keep in my head and reason simply about.

**Comment by devernay** [comment]
Yes, we would all like to live in a linear colorspace world: in this case, colorspace conversion and premultiplication commute, but as soon as the colorspace is non-linear, they dont commute anymore

A and B are linear (we can have bugs and nobody ever notices):
convert_A_to_B(RGB_unpremultiplied) * alpha = convert_A_to_B(RGB_unpremultiplied * alpha)

either A or B is not linear, eg convert_A_to_B(R,G,B) = (R^2.2,G^2.2,B^2.2):
convert_A_to_B(RGB_unpremultiplied) * alpha != convert_A_to_B(RGB_unpremultiplied * alpha)
R^2.2 * alpha != (R*alpha)^2.2

**Comment by devernay** [comment]
maybe OCIO can tell you conversion from A to B is linear, so that it commutes with premult, and thus you don't have to unpremult before / premult after colorspace conversion?

If it doesn't, it probably should! This would save CPU cycles big time!

**Comment by lgritz** [comment]
I'm not sure that it's enough to be linear. The color space transform also needs to have no cross-talk between the channels, I think. Otherwise, there could be some kind of hue shift when comparing the transform alone versus the unpremult-transform-premult, no?

**Comment by lgritz** [comment]
So it looks like the --autocc does the color correct without bracketing with unpremult/premult, so I think the right solution is that it should turn the unpremult/premult back on if the file was read from an unassociated alpha file.

But now I'm all confused again.

Does this mean we're doing the wrong color transform for files with associated alpha? Like, what's the difference between a premultiplied-after-reading PNG versus a premultiplied-in-the-file TIFF?

Does this mean we should always have --autocc do the full unpremult-transform-premult? That's not only super expensive, but it also means we've been doing it wrong and if we fix it, everybody's images are going to change, which is going to be really inconvenient. (And this is used constantly in production, with no complaints. Am I overthinking?)

And now I'm remembering what has always bothered me about this topic. Let's say you have a two image files: A has some non-1.0 alpha, and B is A over opaque black. Those two files differ in alpha (in that B's alpha is 1.0 everywhere) but are exactly the same in their color values in all pixels. Why should those color transform differently?

@michdolan Can you bail me out here and help me reason correctly about this?

**Comment by devernay** [comment]
> I'm not sure that it's enough to be linear. The color space transform also needs to have no cross-talk between the channels, I think. Otherwise, there could be some kind of hue shift when comparing the transform alone versus the unpremult-transform-premult, no?

linear with no bias (ie black transforms to black) is OK, even if there's crosstalk (ie RGB->XYZ): it commutes with multiplication by a scalar (but would not if it were a different scalar for each channel)

**Comment by michdolan** [comment]
When trying to wrap my head around this stuff I often look to Nuke as an example. Nuke's Read node requires you to set a `color space` knob to define the file's input color space (and supports various mechanisms for automatically choosing this for you), which will then be used to convert to the scene-linear working space. So effectively this is the OIIO `--autocc` mode.

By default Nuke will NOT bookend this color space conversion with an unpremult/premult. It is up to the user to enable the `premultiplied` knob telling Nuke that the input image has an associated alpha, at which point it will do the unpremult/premult when converting color.

Similarly most Nuke nodes that manipulate pixels have a `(un)premultiply` knob (again, disabled by default) which will bookend the underlying op with an unpremult and premult.

So by default Nuke doesn't take an associated alpha into account when manipulating color pixels. BUT. users can enable this behavior everywhere if desired, and a studio can override the default to enforce this behavior is they wish.

I find the trickiest aspect of this to be knowing if the input image has an associated alpha. By spec PNG has an unassociated alpha, so you _could_ assume that to be true, though there is nothing stopping a user from writing a PNG with an associated alpha, since as you point out you can't easily determine from inspecting the pixels if the alpha is associated or not.

In the case of OIIO, if you are doing any sort of compositing (specifically merging of layers) are you currently assuming that all inputs have associated alphas? And if so, are you attempting to associate alpha channels on input into OIIO based on each formats spec? For example, always premultiplying a PNG on input, but not doing this for an EXR? I think the answer to that could have some bearing on this decision.

In the end though, I think it could make sense to take the answer to this from Nuke's book. Particularly since changing the `--autocc` behavior will potentially change the output for known inputs in OIIO, I would suggest making this a global and local option. So you could have a single option to tell `oiiotool` to do this bookend unpremult/premult for all color operations, assuming that the user knows this is the right decision given their inputs and how OIIO deals with varying alpha associations on read. Additionally you could potentially support a per-input or per-option flag to enable or disable this behavior, depending on the cost/complexity of implementing that.

The current OIIO ImageBufAlgo color ops support this option already and default to it being false is `oiiotool`, correct? So maybe the right approach here is to similarly add an optional modifier to `--autocc` like `:unpremult=bool` and keep it defaulted to false to maintain previous behavior. 

Even though enabling this option would be more-correct, it's worth noting that Nuke chose not to default this to on. The tricky part will always be in being confident that the input alpha is correctly interpreted with the intended association. File format specs should help with this determination, but as you pointed out, there's nothing stopping a user from getting around it or doing something that could be misinterpreted. So perhaps the safest bet is just to support the option and document the behavior clearly.

**Comment by lgritz** [comment]
Thanks, @michdolan! This is very helpful, as always.

What I really hate about color science is how often it tells me I've been doing something wrong for years.

The current situation and full set of relevant facts, as best as I understand them is:

* ImageBufAlgo::colorconvert() has an optional bool parameter that controls whether unpremult/premult surrounds the color transform, and that defaults to **true**.
* oiiotool --colorconvert has an optional modifier to control the unpremult/premult, and it defaults to **false**.
* oiiotool --autocc is equivalent to --colorconvert, but lacks the unpremult modifier, so it never does the unpremult/premult surrounding the autocorrection.
* OIIO API conventions by default present all input as associated alpha, automatically premultiplying if they know the data in the file is unassociated (as it is always by spec for PNG, though the PNG reader will honor an open() configuration request to keep the data in unassociated form). And oiiotool does not currently turn on this configuration request, meaning that the PNG pixels will be associated by the time they get to oiiotool's color conversion logic.
* For some color transformations, there is no difference between doing the transform directly on RGB versus bracketing by unpremult/premult, but I'm not sure if there's a way of detecting ahead of time which transformations have this property.
* The above point notwithstanding, we believe that it's *always* correct to unpremult/premult surrounding a color transform if you have an image with associated alpha (which is almost all images we encounter in our productions, since OpenEXR says it should always have associated alpha).
* Nuke input nodes give the option to upremult/premult, but it doesn't do it by default. (What's our confidence level that Nuke users know that they should almost always turn it on?)
* PNG is always supposed to be unassociated. OpenEXR is always supposed to be associated. TIFF has metadata that says whether the alpha is associated or unassociated. There is no reliable way to know if a user or upstream script has contravened the conventions or metadata. 

(Pause for everyone to read that and make corrections if there's still something I don't have right.)

So if that's right, what is the best way to move forward?

At the very least, I think the --autocc needs a modifier that lets people specify that the autocc should do the premult/unpremult bracketing. Presumably the default should be false, for the sake of not changing current behavior as well as matching Nuke.

I really don't like that --colorconvert will have different default than --autocc, but I don't know if being consistent there is worth changing the existing default behavior of these two operations, both of which are HEAVILY used in production and quite often changing people's images is much more painful thing than having errors, as long as those errors are subtle and consistent.

I think it's mostly big production users who would notice and be concerned by subtle subtle changes if we switched defaults. But most/all of the nonlinear files they encounter are plates being ingested (no alpha), and by the time they have alpha they are already in a canonical working linear color space. And as we said, many of the color transforms we would do will commute (i.e., give the same pixel values whether or not we do the unpremult/premult). 

So if we were to change defaults to be more correct/consistent, exactly how big is the set of (images with alpha that would transform differently if we unpremult/premult) X (users who would notice/care)?

**Comment by devernay** [comment]
fun fact: I was doing the same error [in my OpenFX plugins](https://github.com/NatronGitHub/Natron/issues/582#issuecomment-792408899), because I handle OCIO conversion myself rather than using OIIO for that:
- I unpremultiply
- the I apply colorspace conversion (eg linear->sRGB)
- then I re-premultiply the sRGB values
- finally, since JPEG doesn't support alpha, we drop alpha, and end up with premultiplied sRGB values, which don't mean much, since sRGB values cannot be multiplied by anything: it is a nonlinear space where any arithmetic is forbidden.

sRGB values of premultiplied pixels (!= "premultiplied sRGB values") are OK, they correspond to merging over solid black before conversion to sRGB.

**Comment by michdolan** [comment]
Thanks for the detailed breakdown @lgritz . That's really helpful to understand. Regarding your concern:

> I really don't like that --colorconvert will have different default than --autocc

From your explanation it seems like they would match, right? Unless I am misunderstanfing.

> oiiotool --colorconvert has an optional modifier to control the unpremult/premult, and it defaults to false.

> --autocc needs a modifier that lets people specify that the autocc should do the premult/unpremult bracketing. Presumably the default should be false

I think the solution of adding the `--autocc:unpremult` option and defaulting to `false` would be effective. It maintains current behavior, preventing surprises, and provides a path forward for those who want to make an improvement to their process. No matter what changes are made, there will always be an opportunity for someone to do the wrong thing, so the best thing is to provide the tools to do the right thing and trust that they will be used properly. It's a confusing subject without question, but I think an explanation in the `--autocc` docs would be clear enough. Just my 2 cents.

**Comment by devernay** [comment]
I think I've come to the conclusion that bracketing *colorspace* changes with unpremult/premult should be avoided:
- either it's a change between *linear* colorspaces, and unpremult/premult commute with the colorspace change, so it's exactly the same operation if you remove these
- or it's a change between two colorspaces, and one of these is nonlinear, in which case premultiplication is meaningless anyway, so the user who wants to convert premultiplied linear to premultiplied srgb or log is doing an error. maybe there could be a warning, but bracketing the colorspace change with unpremult/premult does not help, especially if it's done automatically
- that doesn't mean unpremulti/premult (the famous nuke checkbox) is not useful for other effects, which work on unpremultiplied "true" colors (eg matting or others)

**Comment by lgritz** [comment]
Proposed fix in #2890, in which I add a `unpremult=` optional modifier to `--autocc` as well as to individual `-i` and `-o`. I have not changed any defaults. My current thinking is that changing old behavior, as well as deviating from Nuke defaults, are probably bad ideas. But I would entertain a separate discussion if people wanted to argue that the defaults are wrong and that we should change them next time there is a compatibility-breaking major release. (It does bother me a little that the defaults for `IBA::colorconvert()` and `oiiotool --colorconvert` do not match. But it may be past the point of no return on that.)

With the patch of #2890 (please test), 

    oiiotool --autocc:unpremult=1 ramp.png -o ramp-lg.exr

I believe yields the correct result, and much brighter looking than when you use `--autocc` without the `:unpremult=1`.

@devernay, I don't follow the logic of your last comment. I thought both you and @michdolan had convinced me that it's safest for all color transformations to happen on unassociated color values, and that at worst it might be needlessly expensive for those transformations for which it makes no difference.

@michdolan, can you explain in a way that I'm likely to understand which types of transformations will yield incorrect results if we set the unpremult flag the wrong way? (The grossly nonlinear ones, like sRGB or log based color spaces, are likely to be either plate ingest or output for display, and thus will probably not have any meaningful alpha per se.)

**Comment by devernay** [comment]
Yes maybe I wasn't clear, but the main thing is that any conversion between two linear colorspaces is just a 3x3 matrix conversion (see https://www.colour-science.org:8010/apps/rgb_colourspace_transformation_matrix).

What you are implementing is, supposing R,G,B are premultiplied, to always apply the conversion on unassociated values:
convert(R,G,B,A) = (A * dot(M, [R,G,B] / A) , A)

but you will agree that (that's what I mean by "they commute"):
dot(M,[R,G,B]/A) = 1/A * dot(M,[R,G,B])

so in the end
convert(R,G,B,A) = (dot(M,(R,G,B)) , A)

which means you don't need the premult/unpremult operations. If the OCIO API could tell you whether a colorspace transform is just applying a 3x3 transformation, that would save you the premult/unpremult

**Comment by michdolan** [comment]
Good points @devernay . It's worth noting that in the case of Nuke the "(un)premultiply" knob would generally be used on adjustments performed on linear data. The "premultiplied" knob on the other hand could be used on linear or non-linear inputs, to varying levels of "correctness" perhaps. 

I will add a third case which is not just a matrix transform and also not a conversion to an expressly non-linear encoding. Consider the role of a view transform or "look" applied on linear data. These are creative adjustments performed on linear data that can be non-linear, but are applied in a state where the pixels are still linear, before applying the display encoding. In ACES this would be an LMT, and in OCIO v2 a ViewTransform or Look. In OCIO v2 we now designate these spaces as being display-linear vs. scene-linear since they are in a state in which they can be transformed for various display types, but have the creative intent baked into the pixels.

OCIO v2 can tell you if a processor has cross-talk, or if it is an identity transform, but beyond that would require a much more thorough introspection to make this determination of a transform.

@lgritz to your question specifically, I think @devernay raises a valid point, and honestly something I have probably not thought through well enough. Do you premultiply before or after a non-linear transformation. And the point that a linear alpha should be multiplied by linear RGB data makes sense to me. Though this does raise the question of how PNG handles this. If it's an unassociated alpha, but has non-linear pixels, wouldn't that violate that rule when using the PNG data in a composite? Unless the unpremultiply (to disassociate) was done after applying the non-linear encoding?

@doug-walker any thoughts on this topic from your end? Specifically Larry's question:

> can you explain in a way that I'm likely to understand which types of transformations will yield incorrect results if we set the unpremult flag the wrong way

**Comment by lgritz** [comment]
I think the minimal fix (and maybe the maximum one we can do practically, without unwelcome behavior changes) is what I offer in #2890: just give the option to have --autocc do the unpremult/premult bracketing and call it a day.

I'm 90% convinced that the unpremult dance is what you usually want (and then 10% of the time my brain flips and I bolt awake in a sweat unable to understand why comping over opaque black and transforming should yield different results than transforming the fg element only with partial alpha). I'm still groping for how to explain this all in the docs so that people understand in simple terms when they should turn that option on and when not to.

I was thinking the same as you, @michdolan, which is that in practice most of our transformations are not just a 3x3 matrix, there's always a little bit of nonlinearity (even if just the gentle roll-off at the extremes) even for semantically-almost-linear color spaces.

And you put your finger right on what bugs me about PNG, which is that it not only allows but requires two concepts I don't want to see anywhere near production -- unassociated alpha, and sRGB colors. It seems like a recipe for people and software to do the wrong things and to reason about it all incorrectly, as I have so clearly demonstrated myself.

**Comment by lgritz** [comment]
I merged #2890 into master and will also backport to the next 2.2.x release.

@devernay Do you think that is enough to solve your problems in this case and we can close the issue? Or is there more we can do to help you with your original problem that led you to open this issue?

**Comment by devernay** [comment]
Yes, looks like a correct solution to me!
